### ISSUE TITLE
`return_self_must_use` is reported for types that already have a `#[must_use]` attribute

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust-clippy#8279]
### Summary

Clippy reports `return_self_must_use` for traits in a generic context, even if the return type already has `#[must_use]`.

### Lint Name

return_self_must_use

### Reproducer

I tried this code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=c6a72eb40b979421206de10daf27d85d

```rust
use std::fmt::Display;
use std::error::Error;
pub type BoxError = Box<dyn Error + Send + Sync>;
pub type MyResult<T> = Result<T, BoxError>;

pub trait MyResultExt<T, E> {
    fn log_err(self, msg: &str) -> Self
    where
        E: Display;
}

impl<T> MyResultExt<T, BoxError> for MyResult<T> {
    fn log_err(self, _: &str) -> Self {
        self
    }
}

fn main() {
    Ok(()).log_err("oops");
}
```

I saw this happen:

```
warning: missing `#[must_use]` attribute on a method returning `Self`
 --> src/main.rs:7:5
  |
7 | /     fn log_err(self, msg: &str) -> Self
8 | |     where
9 | |         E: Display;
  | |___________________^
  |
  = note: `#[warn(clippy::return_self_not_must_use)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#return_self_not_must_use
warning: unused `std::result::Result` that must be used
  --> src/main.rs:19:5
   |
19 |     Ok(()).log_err("oops");
   |     ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_must_use)]` on by default
   = note: this `Result` may be an `Err` variant, which should be handled
```

I expected to see this happen: No `return_self_not_must_use` warning, because the compiler already reports `unused_must_use` for this code.

### Version

```text
1.60.0-nightly (2022-01-13 22e491ac7ed454d34669)
```

**Comment by jyn514** [comment]
Hmm, I suppose you can imagine a situation where there's an implementor that doesn't have `#[must_use]`. But in that case I think clippy shouldn't ever warn on the trait method itself, only on the implementing function.
